designs/node-ownership.md

```diff
@@ -0,0 +1,225 @@
+
+#### Provisioner Loop Synchronization
+
+Without in-flight capacity modeling and approximations, Karpenter would have to wait for all instances that it launched to be registered and to report their allocatable capacity before continuing on to the next round of provisioning for pending pods. In practicality, this process of waiting is far too slow (40s+) and requires synchronization between the Karpenter provisioning loop and the node registration.
+
+We can get around waiting and get closer to a better solution here by modeling in-flight capacity in some way (e.g. storing nodes in memory, storing nodes on the cluster, etc.)
+
+#### Node Reconciliation
+
+Karpenter’s ownership model necessitates that it own the VM instances that it launches. We do not want to orphan instances that we launch as the user expects that there should be an easy deletion mechanism for deleting nodes that Karpenter launches (i.e. deleting the Provisioner deletes all VM instances associated with it or deleting a node deletes the VM instance associated with the node).
+
+Without creating the node, we are fully contingent on node object reconciliation to add ownership references and termination finalizers. This is a precarious position to be in since it creates a host of other race conditions that can happen in between the time when Karpenter launches the instance and when the node is created at the api-server. As some examples:
+
+1. What happens if Karpenter controller is deleted in the middle of instance launch but before node creation?
+   1. Answer: the VM instance and node are orphaned
+2. What happens if Karpenter Provisioner is removed in the middle of instance launch but before node creation?
+   1. Answer: the VM instance and node are orphaned
+
+## Current Node Creation Problems
+
+This section describes the primary issues that have arisen from users and from maintainer testing due to Karpenter creating the node object after launching the VM instance in the provisioning loop.
+
+### Synchronization
+
+Node objects currently capture the in-flight details for Karpenter’s scheduler to use (allocatable and capacity details for nodes). As a result, the node object *must* be created by the end of the provisioning loop before provisioning can continue. For AWS, this is not an issue since node names can be resolved based on the EC2 **CreateFleet** response; however, for cloud providers such as Azure, this becomes more of an issue, as node name cannot be resolved from the VM create response but can only be resolved when the kubelet creates the node.
+
+While node launch times might be reduced with time, this means that for some cloudproviders (such as Azure), creating the node ties our inflight logic directly to how long the node takes to launch and register with the api-server. This is a non-starter for our scheduling logic since part of Karpenter’s value proposition is its ability to quickly provision capacity for pending pods.
```

Comment:
On Azure, as long as VM is created directly - rather than via VM Scale Set - one can generate and specify the host name (node name) upfront, as input to Create VM API call. In this mode of operation (which is very likely to be used by Karpenter Azure provider), waiting to find out the name of the node is not a concern.

Reply:
Got it, I can edit this. It seems like an either/or in this case. Either your cloudprovider has to lookup the node name after creation _or_ it has to know the node name before lauch. Looked into it a bit and talked to some of the AWS cloud-controller-manager folks. The TLDR is that this doesn't look like a solution that adequately scales to all cloudproviders and what it comes down to is that we would still create the node before the kubelet has registered it, meaning we are still ending up with strange error messages (like the one mentioned in [this issue](https://github.com/aws/karpenter/issues/2385)) that result from the node being registered before the underlying instance is actually there.

Reply:
Agreed, what we have settled on is relying on the `.spec.providerID` as the mapping between node/machine on the cluster, which is a known mapping (CCM uses this for instance). If node name == machine name == instance name, that's just a benefit for user observability